tools/golang/pkg/encryption/encryption.go

```diff
@@ -0,0 +1,295 @@
+	if err != nil {
+		return "", err
+	}
+	kh, err := insecurecleartextkeyset.Read(&keyset.MemReaderWriter{Keyset: &ks})
+	if err != nil {
+		return "", err
+	}
+	parsedHtml, err := html.Parse(strings.NewReader(htmlStr))
+	if err != nil {
+		return "", err
+	}
+	encryptedSections := getAllEncryptedSections(parsedHtml)
+	if len(encryptedSections) == 0 {
+		return "", errors.New("No encrypted sections found.")
+	}
+	if err = encryptAllSections(parsedHtml, encryptedSections, kh); err != nil {
+		return "", err
+	}
+	encryptedKeys, err := encryptDocumentKey(base64.StdEncoding.EncodeToString(ksEnc), accessRequirement, pubKeys)
+	if err != nil {
+		return "", err
+	}
+	if err = addEncryptedDocumentKeyToHead(encryptedKeys, parsedHtml); err != nil {
+		return "", err
+	}
+	return renderNode(parsedHtml), nil
+}
+
+// Retrieves a Tink public key from the given URL.
+func RetrieveTinkPublicKey(publicKeyUrl string) (tinkpb.Keyset, error) {
+	resp, err := http.Get(publicKeyUrl)
+	if err != nil {
+		return tinkpb.Keyset{}, err
+	}
+	r := keyset.NewJSONReader(resp.Body)
+	ks, err := r.Read()
+	if err != nil {
+		return tinkpb.Keyset{}, err
+	}
+	return *ks, nil
+}
+
+// Generates a new AES-GCM key.
+func generateNewAesGcmKey(km registry.KeyManager) ([]byte, error) {
+	p, err := proto.Marshal(&gcmpb.AesGcmKeyFormat{KeySize: aesGCMKeySize})
+	if err != nil {
+		return nil, err
+	}
+	m, err := km.NewKey(p)
+	if err != nil {
+		return nil, err
+	}
+	return proto.Marshal(m)
+}
+
+// Creates an AES-GCM Keyset using the input key.
+// Example output proto:
+// 		primary_key_id: 1
+// 		key: <
+// 			key_data: <
+//   			type_url: "type.googleapis.com/google.crypto.tink.AesGcmKey"
+//   			value: "\032\020\355\323'\277\341\241u\020w\322\177\207\357\374\301/"
+//   			key_material_type: SYMMETRIC
+// 			>
+// 			status: ENABLED
+// 			key_id: 1
+// 			output_prefix_type: TINK
+// 		>
+func createAesGcmKeyset(key []byte) tinkpb.Keyset {
+	keyData := tinkpb.KeyData{
+		KeyMaterialType: tinkpb.KeyData_SYMMETRIC,
+		TypeUrl:         aesGCMKeyURL,
+		Value:           key,
+	}
+	keys := []*tinkpb.Keyset_Key{
+		&tinkpb.Keyset_Key{
+			KeyData:          &keyData,
+			Status:           tinkpb.KeyStatusType_ENABLED,
+			KeyId:            1,
+			OutputPrefixType: tinkpb.OutputPrefixType_TINK,
+		},
+	}
+	return tinkpb.Keyset{
+		PrimaryKeyId: 1,
+		Key:          keys,
+	}
+}
+
+// Retrieves all encrypted content sections from the parsed HTML tree.
+func getAllEncryptedSections(parsedHtml *html.Node) []*html.Node {
+	for n := parsedHtml.FirstChild; n != nil; n = n.NextSibling {
+		if (n.DataAtom == atom.Html) && (len(n.Attr) != 0) {
+			for bn := n.FirstChild; bn != nil; bn = bn.NextSibling {
+				if bn.DataAtom == atom.Body {
+					return getEncryptedSectionsDfs(bn)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// Searches for all <section subscriptions-section="content" encrypted> nodes and returns them.
+func getEncryptedSectionsDfs(bodyNode *html.Node) []*html.Node {
```

Comment:
nit: this should be DFS. Yes, annoying. literally every acronym should be constant case.